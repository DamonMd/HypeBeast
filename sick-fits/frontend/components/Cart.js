import React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import SickButton from "./styles/SickButton";
import CartStyle from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";

const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Cart = () => {
  return (
    <Mutation mutation={TOGGLE_CART_MUTATION}>
      {toggleCart => (
        <Query query={LOCAL_STATE_QUERY}>
          {({ data }) => (
            <CartStyle open={data.cartOpen} onClick={toggleCart}>
              <header>
                <CloseButton title="close">&times;</CloseButton>
                <Supreme>Your Cart</Supreme>
                <p> you have ....... items in your cart</p>
              </header>
              <footer>
                <p>$92.10</p>
                <SickButton>Checkout</SickButton>
              </footer>
            </CartStyle>
          )}
        </Query>
      )}
    </Mutation>
  );
};

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };
