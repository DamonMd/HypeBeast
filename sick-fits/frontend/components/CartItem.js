import React from "react";
import formatMoney from "../lib/formatMoney";
import styled from "styled-components";
import PropTypes from "prop-types";
import RemoveFromCart from "./RemoveFromCart";

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CartItem = props => {
  // 1. check if item exist on the site still
  if (!props.cartItem.item) {
    return (
      <CartItemStyles>
        <p>This product has been removed from the site</p>
        <RemoveFromCart id={props.cartItem.id} />
      </CartItemStyles>
    );
  }
  const { image, price, title, id, description } = props.cartItem.item;

  return (
    <CartItemStyles>
      <img width="100" src={image} alt={title} />
      <div className="cart-item-details">
        <h3>{title}</h3>
        <p>
          {formatMoney(price * props.cartItem.quantity)} {" - "}{" "}
          <em>
            {props.cartItem.quantity} &times; {formatMoney(price)}
          </em>
        </p>
      </div>
      <RemoveFromCart id={props.cartItem.id} />
    </CartItemStyles>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired
};

export default CartItem;
