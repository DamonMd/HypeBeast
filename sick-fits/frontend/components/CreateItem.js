import React, { Component } from "react";
import Form from "./styles/Form";
import gql from "graphql-tag";
import Router from "next/router";
import { Mutation } from "react-apollo";
import Error from "./ErrorMessage";

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: "",
    description: "",
    price: 0,
    image: "",
    largeImage: ""
  };

  handleChange = e => {
    const { name, value, type } = e.target;
    const val = type === "number" ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  uploadFile = async e => {
    const files = e.target.files;
    const data = new FormData();
    data.append("file", files[0]);
    //cloudinary config settings
    data.append("upload_preset", "hypebeast");
    const res = await fetch(
      "https://api.cloudinary.com/v1_1/dokwu1eus/image/upload",
      {
        method: "POST",
        body: data
      }
    );
    const file = await res.json();
    // const file = res;
    console.log(file);
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    });
  };

  render() {
    const { title, price, description, image } = this.state;
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              e.preventDefault();
              //TODO: add in check to make sure cloudinary has returned a response
              //call mutation
              const response = await createItem();
              //re-route to single item page
              console.log(response);
              Router.push({
                pathname: "/item",
                query: { item: response.data.createItem.id }
              });
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="title">Image</label>
              <input
                type="file"
                id="file"
                name="file"
                placeholder="Upload an image"
                onChange={this.uploadFile}
              />
              {this.state.image && (
                <img src={image} width="200" alt="Upload Preview" />
              )}
              <label htmlFor="title">Title</label>
              <input
                type="text"
                name="title"
                value={title}
                placeholder="Title"
                onChange={this.handleChange}
              />
              <label htmlFor="price">Price</label>
              <input
                type="number"
                name="price"
                value={price}
                onChange={this.handleChange}
              />
              <label htmlFor="description">Description</label>
              <textarea
                type="text"
                name="description"
                value={description}
                placeholder="Enter a description"
                onChange={this.handleChange}
              />
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
