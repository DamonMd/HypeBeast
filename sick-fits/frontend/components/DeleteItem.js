import React, { Component } from "react";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";
import { ALL_ITEMS_QUERY } from "./Items";

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {
  update = (cache, payload) => {
    //manually update the cache on client to match the server
    //read the cache for the items we want
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    console.log(data);
    console.log(payload);
    // filter deleted items out of the page
    data.items = data.items.filter(
      item => item.id !== payload.data.deleteItem.id
    );
    console.log("now the items areee....", data);
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data: data });
  };

  render() {
    const { id } = this.props;
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: id }}
        update={this.update}
      >
        {(deleteItem, { error }) => (
          <button
            onClick={() => {
              if (confirm("you sure about deleting this?")) {
                deleteItem();
              }
            }}
          >
            {this.props.children}
          </button>
        )}
      </Mutation>
    );
  }
}

export default DeleteItem;
