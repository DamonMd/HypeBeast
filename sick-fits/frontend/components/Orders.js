import React, { Component } from "react";
import Error from "./ErrorMessage";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import PropTypes from "prop-types";
import formatMoney from "../lib/formatMoney";
import Link from "next/link";
import styled from "styled-components";
import OrderItemStyles from "./styles/OrderItemStyles";

const ORDERS_QUERY = gql`
  query ORDERS_QUERY {
    orders {
      id
      total
      items {
        id
        price
        image
        quantity
        description
        title
      }
    }
  }
`;

const OrderUl = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

class Orders extends Component {
  render() {
    return (
      <div>
        <Query query={ORDERS_QUERY}>
          {({ data, loading, error }) => {
            if (error) return <Error error={error} />;
            if (loading) return <p>loading....</p>;
            console.log("data from order query", data);
            const orders = data.orders;
            return (
              <div>
                <h2>You have {orders.length} orders</h2>
                <OrderUl>
                  {orders.map(order => {
                    return (
                      <OrderItemStyles key={order.id}>
                        <Link
                          href={{
                            pathname: "/order",
                            query: { id: order.id }
                          }}
                        >
                          <a>
                            <div className="order-meta">
                              <p>
                                {order.items.reduce(
                                  (a, b) => a + b.quantity,
                                  0
                                )}{" "}
                                Items
                              </p>
                              <p>{formatMoney(order.total)}</p>
                            </div>
                            <div className="images">
                              {order.items.map(item => {
                                return (
                                  <img
                                    key={item.id}
                                    src={item.image}
                                    alt={item.description}
                                  />
                                );
                              })}
                            </div>
                          </a>
                        </Link>
                      </OrderItemStyles>
                    );
                  })}
                </OrderUl>
              </div>
            );
          }}
        </Query>
      </div>
    );
  }
}

export default Orders;
