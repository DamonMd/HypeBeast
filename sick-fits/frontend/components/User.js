import { Query } from "react-apollo";
import gql from "graphql-tag";
import PropTypes from "prop-types";
import SickButton from "./styles/SickButton";

const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`;

const User = props => {
  return (
    <Query query={CURRENT_USER_QUERY}>
      {payload => props.children(payload)}
    </Query>
  );
};

const UserRights = props => {
  console.log("p", props);
  return (
    <tr>
      <td>{props.user.name}</td>
      <td>{props.user.email}</td>
      {props.permissions.map(p => (
        <td>
          <input type="checkbox" />
        </td>
      ))}
      <td>
        <SickButton>UPDATE</SickButton>
      </td>
      {/* {props.permissions.map(p => {
        return (
          <td>
            <label htmlFor={`${user.id}-permission-${p}`} />
            <input type="checkbox" />
          </td>
        );
      })} */}
    </tr>
  );
};

User.propTypes = {
  children: PropTypes.func.isRequired
};

export default User;
export { CURRENT_USER_QUERY };
export { UserRights };
