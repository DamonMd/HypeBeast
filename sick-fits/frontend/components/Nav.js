import React from "react";
import Link from "next/link";
import { Mutation } from "react-apollo";
import NavStyles from "./styles/NavStyles";
import User from "./User";
import Signout from "./Signout";
import { TOGGLE_CART_MUTATION } from "./Cart";
import CartCount from "./CartCount";

const Nav = () => {
  return (
    <User>
      {({ data: { me } }) => (
        <NavStyles data-test="nav">
          <Link href="/items">
            <a>shop</a>
          </Link>
          {me && (
            <>
              <Link href="/sell">
                <a>Sell</a>
              </Link>
              <Link href="/orders">
                <a>orders</a>
              </Link>
              <Link href="/me">
                <a>account</a>
              </Link>
              <Signout />
              <Mutation mutation={TOGGLE_CART_MUTATION}>
                {toggleCart => (
                  <button onClick={toggleCart}>
                    My Cart{" "}
                    <CartCount
                      count={me.cart.reduce((total, cart) => {
                        return total + cart.quantity;
                      }, 0)}
                    />
                  </button>
                )}
              </Mutation>
            </>
          )}
          {!me && (
            <Link href="/signup">
              <a>Sign In</a>
            </Link>
          )}
        </NavStyles>
      )}
    </User>
  );
};

export default Nav;
