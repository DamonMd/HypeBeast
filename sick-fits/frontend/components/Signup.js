import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Form from "./styles/Form";
import Error from "./ErrorMessage";
import { CURRENT_USER_QUERY } from "./User";

const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $name: String!
    $email: String!
    $password: String!
  ) {
    signup(name: $name, email: $email, password: $password) {
      id
      email
    }
  }
`;

class Signup extends Component {
  state = {
    name: "",
    password: "",
    email: ""
  };

  handleChange = e => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  submitForm = e => {
    e.preventDefault();
    console.log("idk");
  };

  render() {
    return (
      <Mutation
        mutation={SIGNUP_MUTATION}
        variables={this.state}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {(signup, { error, loading }) => {
          return (
            <Form
              data-test="Signup"
              method="post"
              onSubmit={async e => {
                e.preventDefault();
                const res = await signup();
              }}
            >
              <fieldset disabled={loading} aria-disabled={loading}>
                <h2>Sign up for an account</h2>
                <Error error={error} />
                <label>
                  Name
                  <input
                    name="name"
                    type="text"
                    placeholder="name"
                    value={this.state.name}
                    onChange={this.handleChange}
                  />
                </label>

                <label>
                  Email
                  <input
                    name="email"
                    type="email"
                    placeholder="email"
                    value={this.state.email}
                    onChange={this.handleChange}
                  />
                </label>

                <label>
                  Password
                  <input
                    name="password"
                    type="password"
                    placeholder="password"
                    value={this.state.password}
                    onChange={this.handleChange}
                  />
                </label>

                <button type="submit">Sign Up!</button>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }
}

export default Signup;
export { SIGNUP_MUTATION };
