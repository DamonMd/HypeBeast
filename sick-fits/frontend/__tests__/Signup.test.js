import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Signup, { SIGNUP_MUTATION } from "../components/Signup";
import { CURRENT_USER_QUERY } from "../components/User";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeUser } from "../lib/testUtils";
import { ApolloConsumer } from "react-apollo";

function type(wrapper, name, value) {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
}

const me = fakeUser();
const mocks = [
  {
    // sign up mutation
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        name: me.name,
        password: "wes",
        email: me.email
      }
    },
    result: {
      signup: {
        __typename: "User",
        id: "abc123",
        email: me.email,
        name: me.name
      }
    }
  },
  // current user mock
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: { data: { me } }
  }
];

describe("<Signup/>", () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider>
        <Signup />
      </MockedProvider>
    );
    const form = wrapper.find('form[data-test="Signup"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it("calls the mutation correctly", async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <Signup />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    type(wrapper, "name", me.name);
    type(wrapper, "email", me.email);
    type(wrapper, "password", "wes");
    wrapper.update();
    // wrapper.find("form").simulate("submit");
    // await wait();
    //query the user out of the apollo client
    const user = await apolloClient.query({ query: CURRENT_USER_QUERY });
  });
});
