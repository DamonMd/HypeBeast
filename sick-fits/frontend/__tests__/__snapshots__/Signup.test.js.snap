// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 dMniLL"
  data-test="Signup"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-disabled={false}
    disabled={false}
  >
    <h2>
      Sign up for an account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label>
      Name
      <input
        name="name"
        onChange={[Function]}
        placeholder="name"
        type="text"
        value=""
      />
    </label>
    <label>
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="email"
        type="email"
        value=""
      />
    </label>
    <label>
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up!
    </button>
  </fieldset>
</form>
`;
